### PR TITLE
[BrowserKit] Emulate back/forward browser navigation

### PR DESCRIPTION
**Pull request by e-moe**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | yes
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #22336
| License       | MIT
| Doc PR        | CHANGELOG.md updated

Hi all, please review this code for emulating back/forward browser navigation (skip redirects). If code is ok I will add tests and docs

### CODE REVIEW

**Comment by sstok** [comment]
As this is a new feature please rebase this on the master and change the pull-request target. Also this is missing some tests.

Status: needs work

**Comment by e-moe** [comment]
@sstok , thanks for your comment. PR target is changed to master.

as for tests - I'm waiting for implementation review/approval. if it's done correctly I will add missed tests

File: src/Symfony/Component/BrowserKit/Client.php
Comment (stof) [inline src/Symfony/Component/BrowserKit/Client.php]: you should use the hash as key rather than value. the array is used to check whether it contains a given hash, and `O(1)` is better than than `O(n)`

File: src/Symfony/Component/BrowserKit/Client.php
Comment (stof) [inline src/Symfony/Component/BrowserKit/Client.php]: please use private visibility for any new property. We don't want to make them extension points yet

File: src/Symfony/Component/BrowserKit/Client.php
Comment (stof) [inline src/Symfony/Component/BrowserKit/Client.php]: what if we reach the beginning of the history ?

Comment (e-moe) [inline src/Symfony/Component/BrowserKit/Client.php]: History object will throw `\LogicException('You are already on the first page.');`

File: src/Symfony/Component/BrowserKit/Client.php
Comment (e-moe) [inline src/Symfony/Component/BrowserKit/Client.php]: done

File: src/Symfony/Component/BrowserKit/Client.php
Comment (stof) [inline src/Symfony/Component/BrowserKit/Client.php]: then fine

File: src/Symfony/Component/BrowserKit/Client.php
Comment (e-moe) [inline src/Symfony/Component/BrowserKit/Client.php]: done

File: src/Symfony/Component/BrowserKit/Client.php
Comment (e-moe) [inline src/Symfony/Component/BrowserKit/Client.php]: the behavior will be the same

**Comment by javiereguiluz** [comment]
I'd like to ask if this is something that browsers do contrary to the standard ... or if this is the behavior proposed by the related standard. That way we'll know if this is a new feature or a bug. Thanks!

**Comment by e-moe** [comment]
@javiereguiluz , maybe we can check this document: https://www.w3.org/TR/html5/browsers.html#history

**Comment by curry684** [comment]
Assuming the BrowserKit client should conform to that standard it's a bug, as [`history.back()`](https://www.w3.org/TR/html5/browsers.html#history) is defined to:

> Goes back one step in the joint session history.
>
> If there is no previous page, does nothing.

[Joint session history](https://www.w3.org/TR/html5/browsers.html#joint-session-history) is defined as (emphasis moved):

> The joint session history of a top-level browsing context is the union of all the session histories of all browsing contexts of **all the fully active Document objects that share that top-level browsing context**, with all the entries that are current entries in their respective session histories removed except for the current entry of the joint session history.

As a redirect is no fully active Document object navigation should bypass it, hence it's a bug.

But also a BC breaking bug as there may be tons of tests out there explicitly going back twice in history to circumvent this behavior.

**Comment by e-moe** [comment]
@curry684 , thanks for you findings! Sine this is a bugfix should I rebase this PR to 2.7 branch?

**Comment by fabpot** [comment]
As it might break existing tests, I think this needs to be merged in master, even if this is a bug fix.

File: src/Symfony/Component/BrowserKit/Client.php
Comment (fabpot) [inline src/Symfony/Component/BrowserKit/Client.php]: I would not make it configurable as browsers don't make this configurable either. I understand that you probably want to be able to keep BC here. But I don't think that this is a good idea. The current behavior is wrong and we should fix it. What needs to be done though is to add a note in the CHANGELOG/UPGRADE files to document this change.

**Comment by curry684** [comment]
Given the possibly severe BC breaks I'd also recommend considering this a functionality change, hence master.

File: src/Symfony/Component/BrowserKit/Client.php
Comment (e-moe) [inline src/Symfony/Component/BrowserKit/Client.php]: done

**Comment by e-moe** [comment]
Tests and docs added, comments applied.

Also, I changed `spl_object_hash` to `serialize` due to `clone` usage in History object: hashes was new all the time (see https://github.com/symfony/symfony/blob/master/src/Symfony/Component/BrowserKit/History.php#L68)

File: src/Symfony/Component/BrowserKit/Client.php
Comment (staabm) [inline src/Symfony/Component/BrowserKit/Client.php]: did you consider `spl_object_hash` instead of serialize? It should be way faster and less memory intensive.

http://php.net/manual/en/function.spl-object-hash.php

Comment (stloyd) [inline src/Symfony/Component/BrowserKit/Client.php]: See comment above done by @e-moe explaining why it was changed.

**Comment by willemstuursma** [comment]
Anything I can do to make this go forward?

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @e-moe.

**Pull request closed without merging**
